apollo-router/src/services/subgraph_service.rs

```diff
@@ -597,104 +593,205 @@ async fn call_http(
+    // 3. If the response type is `application/json` and status is not 2xx and the body is not valid grapqhql then parse errors will be suppressed.
 
-        // Keep our parts, we'll need them later
-        let (parts, body) = response.into_parts();
-        if display_headers {
-            tracing::info!(
-                        http.response.headers = ?parts.headers, apollo.subgraph.name = %service_name, "Response headers from subgraph {service_name:?}"
-                    );
-        }
-        if let Some(content_type) = parts.headers.get(header::CONTENT_TYPE) {
-            if let Ok(content_type_str) = content_type.to_str() {
-                // Using .contains because sometimes we could have charset included (example: "application/json; charset=utf-8")
-                if !content_type_str.contains(APPLICATION_JSON.essence_str())
-                    && !content_type_str.contains(GRAPHQL_JSON_RESPONSE_HEADER_VALUE)
-                {
-                    cloned_context.leave_active_request();
+    let display_headers = context.contains_key(LOGGING_DISPLAY_HEADERS);
+    let display_body = context.contains_key(LOGGING_DISPLAY_BODY);
 
-                    return if !parts.status.is_success() {
+    // Print out the debug for the request
+    if display_headers {
+        tracing::info!(http.request.headers = ?request.headers(), apollo.subgraph.name = %service_name, "Request headers to subgraph {service_name:?}");
+    }
+    if display_body {
+        tracing::info!(http.request.body = ?request.body(), apollo.subgraph.name = %service_name, "Request body to subgraph {service_name:?}");
+    }
 
-                        Err(BoxError::from(FetchError::SubrequestHttpError {
-                            service: service_name.clone(),
-                            status_code: Some(parts.status.as_u16()),
-                            reason: format!(
-                                "{}: {}",
-                                parts.status.as_str(),
-                                parts.status.canonical_reason().unwrap_or("Unknown")
-                            ),
-                        }))
-                    } else {
-                        Err(BoxError::from(FetchError::SubrequestHttpError {
-                            status_code: Some(parts.status.as_u16()),
-                            service: service_name.clone(),
-                            reason: format!("subgraph didn't return JSON (expected content-type: {} or content-type: {GRAPHQL_JSON_RESPONSE_HEADER_VALUE}; found content-type: {content_type:?})", APPLICATION_JSON.essence_str()),
-                        }))
-                    };
-                }
-            }
-        }
+    // Perform the actual fetch. This
+    context.enter_active_request();
+    let (parts, body) = perform_fetch(client, service_name, request)
+        .instrument(subgraph_req_span)
+        .await
+        .map(|r| {
+            context.leave_active_request();
```

Comment:
`leave_active_request` should be called on the error side too, otherwise it will count indefinitely.
I'll open a separate PR to move that accounting to a guard-like behaviour that stops counting when dropped, that will prevent some issues

Reply:
[397a5b4](https://github.com/apollographql/router/pull/3328/commits/397a5b40ed5214987667ae4d2cc78f8b56685c6a)